eclipse/che
# [Preview do not merge] UserDevfile API

### What does this PR do?
Preview of UserDevfile API
<img width="1024" alt="Знімок екрана 2020-06-23 о 16 37 44" src="https://user-images.githubusercontent.com/1614429/85417548-06024c80-b570-11ea-87fa-a2dc5f99e760.png">


 - [x] Model
 - [x]  Dao
 - [x]  Dao test
 - [x] UserDevfile Service 
 - [x] UserDevfile Service Test

### QA
    - Q: What is it?
    - A: This is a set of REST APIs that allows saving/edit/delete/search devfiles
    - Q: Are there any plans for sharing, using them in factories?
    - A: Yes in the next releases.
    - Q: Is This production-ready?
    - A: No, this is a technical preview.   
    - Q: Why the service path  is `/userdevfile` instead of `/devfile`
    - A: We already have `/devfile` REST API where the main `GET` method is used to get `devfile 1.0 ` schema json. I don't think we should change it. 
    - Q: Why do we need UserDevfile model object, we already have Devfile?
    - A: By analog with Workspace model object we need `id` to be able to uniquely identify persisted objects.
    - Q: Why JoinColumn was used in UserDevfileImpl?
    - A: It doesn't require additional filed in `Devfile` table or additional `table`. It might happen because DevfileImpl is used now it two places: WorkspaceImpl and UserDevfileImpl
    - Q: Is there any known issues with persistence.
    - Y:  Yes. `generateName` is not persisted. An alternative could be is to have separate UserDevfileImpl+ MetadataImpl.
    - Q: Why filtering restricted by `devfile.metadata.name` in JpaUserDevfileDao?
    - A: The idea is to identify during review necessary filtering and order fields and after that create DB indexes for them and make sure we can execute queries efficiently. 
    - Q: Is this service would be available in multi-user mode?
    - A: Eventually yes. But not in this iteration. Before the merge, I'm going to restrict usage for single-user only. The reason for that is that we don't have any authorization mechanism available at this moment.

### What issues does this PR fix or reference?
 https://github.com/eclipse/che/issues/16981
